Replace old method of registration with global registration in robottelo

shweta83 opened the pull request:
### Problem Statement
Robottelo is still using old method of registration using katello-ca certs which is deprecated.

### Solution
With this PR, all the instances of old method of registration will be replaced with global registration.

### Related Issues


<!-- ### PRT test Cases example
trigger: test-robottelo
pytest: tests/foreman/ui/test_contenthost.py -k 'test_syspurpose_mismatched'
-->
<!--
PRT usage reference link: https://github.com/SatelliteQE/robottelo/wiki/Robottelo-Pull-Request-Testing-(PRT)-Process#usage-examples
-->

vsedmik commented on tests/foreman/ui/test_activationkey.py:
```diff
@@ -1144,7 +1145,7 @@ def test_positive_delete_manifest(session, function_entitlement_manifest_org, ta
 @pytest.mark.tier3
 @pytest.mark.skipif((not settings.robottelo.repos_hosting_url), reason='Missing repos_hosting_url')
-def test_positive_ak_with_custom_product_on_rhel6(session, rhel6_contenthost, target_sat):
+def test_positive_ak_with_custom_product_on_rhel6(session, rhel_contenthost, target_sat):
```
The supportability question here.

shweta83 replied:
Fixed it. Thanks.

vsedmik commented on tests/foreman/ui/test_reporttemplates.py:
```diff
@@ -515,9 +515,9 @@ def test_positive_gen_entitlements_reports_multiple_formats(
-    client.install_katello_ca(target_sat)
     module_org, ak = setup_content
-    client.register_contenthost(module_org.label, ak.name)
+    result = client.register(module_org, None, [ak.name], target_sat)
```
`[]`

vsedmik commented on tests/foreman/ui/test_reporttemplates.py:
```diff
@@ -597,8 +597,8 @@ def test_positive_generate_all_installed_packages_report(
     client = rhel_contenthost
-    client.install_katello_ca(target_sat)
-    client.register_contenthost(org.label, ak.name)
+    result = client.register(org, None, [ak.name], target_sat)
```
`[]`

vsedmik commented on tests/foreman/ui/test_subscription.py:
```diff
@@ -518,8 +518,8 @@ def test_positive_candlepin_events_processed_by_STOMP(
     ).create()
-    rhel7_contenthost.install_katello_ca(target_sat)
-    rhel7_contenthost.register_contenthost(org.name, ak.name)
+    result = rhel7_contenthost.register(org, None, [ak.name], target_sat)
```
`[]`

vsedmik reviewed (commented):
Thank you @shweta83 for the refactor! Codewisely looks good, I have just put a few comments / questions for a discussion where I wasn't sure about the intention or context.

Gauravtalreja1 commented on tests/foreman/ui/test_activationkey.py:
```diff
@@ -1144,7 +1145,7 @@ def test_positive_delete_manifest(session, function_entitlement_manifest_org, ta
 @pytest.mark.tier3
 @pytest.mark.skipif((not settings.robottelo.repos_hosting_url), reason='Missing repos_hosting_url')
-def test_positive_ak_with_custom_product_on_rhel6(session, rhel6_contenthost, target_sat):
+def test_positive_ak_with_custom_product_on_rhel6(session, rhel_contenthost, target_sat):
```
As you're replacing rhel6_contenthost with rhel_contenthost, and this test looks specific to rhel6, so can you add missing `rhel_ver_match` or `rhel_ver_list` marks on the test for el6?

shweta83 replied:
Yes. I missed adding it. Thanks for pointing.

shweta83 commented on tests/foreman/cli/test_host.py:
```diff
@@ -1909,9 +1904,8 @@ def test_positive_attach(
-    host_subscription_client.register_contenthost(
-        module_org.name, activation_key=module_ak_with_cv.name
-    )
+    result = host_subscription_client.register(module_org, None, module_ak_with_cv.name, target_sat)
```
Thanks for pointing out. I will update it.

ogajduse replied:
Running `host_subscription_client.register` then is not needed, since `host_subscription_client` already comes subscribed, see the `host_subscription_client` fixture.

Can you either
* use `rhel7_contenthost` directly and remove the fixture or
* remove all occurrences of  `host_subscription_client.register`?

shweta83 commented on tests/foreman/api/test_subscription.py:
```diff
@@ -229,8 +229,13 @@ def test_positive_subscription_status_disabled(
     """
-    rhel_contenthost.install_katello_ca(target_sat)
-    rhel_contenthost.register_contenthost(module_sca_manifest_org.label, module_ak.name)
+    command = target_sat.api.RegistrationCommand(
+        organization=module_sca_manifest_org,
+        activation_keys=[module_ak.name],
+        insecure=True,
+    ).create()
+    result = rhel_contenthost.execute(command)
```
I preferred registration through API in API test cases. That's why I used this.

shweta83 commented on tests/foreman/ui/test_activationkey.py:
```diff
@@ -931,7 +929,7 @@ def test_negative_usage_limit(session, module_org, target_sat):
 @pytest.mark.upgrade
 @pytest.mark.skipif((not settings.robottelo.repos_hosting_url), reason='Missing repos_hosting_url')
-def test_positive_add_multiple_aks_to_system(session, module_org, rhel6_contenthost, target_sat):
+def test_positive_add_multiple_aks_to_system(session, module_org, rhel_contenthost, target_sat):
```
Do you think we should test it for RHEL6?

vsedmik replied:
Honestly, I hope to get rhel6 out of supportability soon.

shweta83 replied:
Yeah me too. I raised PR for it https://github.com/SatelliteQE/robottelo/pull/15012. Let's see if we can get rid of it soon

shweta83 commented on tests/foreman/ui/test_activationkey.py:
```diff
@@ -839,21 +839,20 @@ def test_positive_add_host(session, module_org, rhel6_contenthost, target_sat):
 @pytest.mark.skip_if_not_set('clients')
 @pytest.mark.tier3
-def test_positive_delete_with_system(session, rhel6_contenthost, target_sat):
+def test_positive_delete_with_system(session, rhel_contenthost, target_sat):
```
Should we support this for RHEL6? This should work for any other RHEL version.

shweta83 commented on tests/foreman/ui/test_activationkey.py:
```diff
@@ -819,7 +819,7 @@ def test_positive_add_docker_repo_ccv(session, module_org, module_target_sat):
 @pytest.mark.skip_if_not_set('clients')
 @pytest.mark.tier3
-def test_positive_add_host(session, module_org, rhel6_contenthost, target_sat):
+def test_positive_add_host(session, module_org, rhel_contenthost, target_sat):
```
My intention was to run for all supported rhel  versions but please share your thoughts too.

pondrejk reviewed (approved):
looked through the endeavour's components, no issues found

ogajduse commented on tests/foreman/cli/test_host.py:
```diff
@@ -2031,10 +2023,11 @@ def test_negative_without_attach_with_lce(
-        environment=function_lce,
-        organization=function_org,
-    ).create()
+    # ak = target_sat.api.ActivationKey(
+    #     environment=function_lce,
+    #     organization=function_org,
+    #     content_view
+    # ).create()
```
Looking at other comments getting resolved, it feels that this one is sort of forgotten. Can you, please, address it?

ogajduse commented on tests/foreman/cli/test_host.py:
```diff
@@ -2004,6 +1994,7 @@ def test_negative_without_attach(
     )
     host = target_sat.cli.Host.info({'name': host_subscription_client.hostname})
+
     host_subscription_client.register_contenthost(
```
Did you leave this one unchanged intentionally?

ogajduse reviewed (approved):
Nice work, Shweta! Thank you for taking the time to update our tests to use the conversion to Global Registration.
I have left some comments some of which point out to forgotten arguments in function calls, minor improvements, and one of them asking to create a helper function.

ogajduse commented:
Not all my comments were addressed, can you please go through them?
pre-commit is failing, please fix it.
And also the PR has two conflicts.

shweta83 commented:
> Not all my comments were addressed, can you please go through them? pre-commit is failing, please fix it. And also the PR has two conflicts.

I haven't addressed all the comments yet. I was discussing an issue with devs so I am still working on addressing it. I will request your review once I will be done with all.

ogajduse reviewed (approved):
ACK, one minor comment left.

Griffin-Sullivan commented on tests/foreman/api/test_subscription.py:
```diff
@@ -413,8 +430,14 @@ def test_positive_expired_SCA_cert_handling(module_sca_manifest_org, rhel7_conte
     # registration of a host with an expired SCA cert
-    rhel7_contenthost.register_contenthost(module_sca_manifest_org.label, ak.name)
-    assert rhel7_contenthost.subscribed
+    result = rhel_contenthost.api_register(
+        target_sat,
+        organization=module_sca_manifest_org,
+        activation_keys=[ak.name],
+        force=True,
+    )
```
I feel like you could add in the comment above `# re-registering the host (using force=True)......`

shweta83 replied:
I think re-registering implies that but no harm in explicitly mentioning it.

ogajduse replied:
Oh, I must have missed the re-registration. Thanks for updating it anyway.

Griffin-Sullivan commented on robottelo/hosts.py:
```diff
@@ -798,6 +798,17 @@ def register(
         cmd = target.satellite.cli.HostRegistration.generate_command(options)
         return self.execute(cmd.strip('\n'))
 
+    def api_register(self, target, **kwargs):
+        """Register a content host using the API.
```
I feel like we could either name this something related to global registration like `def globally_register` or leave the name as is and update the description to mention global registration like `Register a content host using Global Registration.`

Griffin-Sullivan replied:
Yeah it's not a great name that's why I thought at least the description :smile:

vsedmik replied:
+1 to keep the name and extend docstring

sambible replied:
+1 to this. Tying this in explicitly to global registration should make it more clear why this is the method being used going forward.

Griffin-Sullivan reviewed (approved):
Nothing here should impact Team Platform. Added some comments based on my opinion. Thanks for adding the changes!

shweta83 commented on tests/foreman/cli/test_host.py:
```diff
@@ -2048,19 +2041,17 @@ def test_negative_without_attach_with_lce(
-        auto_attach=False,
+    result = host_subscription_client.register(
+        function_org, None, function_ak_with_cv.name, target_sat
     )
```
global registration template doesn't take `lifecycle environment`. It take whatever lce assigned to `activation_key`.

ogajduse replied:
ACK, let's go with the proposed implementation
